doc/src/documentation_style_guide.rst

```diff
@@ -0,0 +1,1019 @@
+<https://github.com/sympy/sympy/wiki/Development-workflow>`_ guide.
+
+Best Practices for Writing Documentation
+========================================
+
+Please follow these formatting, style, and tone preferences when writing
+documentation.
+
+Formatting Preferences
+----------------------
+
+**Math vs. Code**
+
+**Math:** text that is surrounded by dollar signs $ _ $ will be rendered as
+LaTeX math. Any text that is meant to appear as LaTeX math should be written as
+$math$. In the HTML version of the docs, MathJax will render the math.
+
+**Example**
+
+::
+
+  The Bessel $J$ function of order $\nu$ is defined to be the function
+  satisfying Bessel’s differential equation.
+
+**Code:** text that should be printed verbatim, such as code, should be
+surrounded by a set of double backticks ``like this``. (WORK IN PROGRESS: change
+single backticks to render as code via issue `#13519
+<https://github.com/sympy/sympy/issues/13519>`_.)
+
+**Example**
+
+::
+
+  To use this class, define the ``_rewrite`` and ``_expand`` methods.
+
+Sometimes a variable will be the same in both math and code, and can even
+appear in the same paragraph, making it difficult to know if it should be
+formatted as math or code. If the sentence in question is discussing
+mathematics, then LaTeX should be used, but if the sentence is discussing the
+SymPy implementation specifically, then code should be used.
+
+**Example**
+
+::
+
+  def stirling(n, k, d=None, kind=2, signed=False):
+      """
+      ...
+
+      The first kind of Stirling number counts the number of permutations of
+      ``n`` distinct items that have ``k`` cycles... This counts the ways to
+      partition $n$ consecutive integers into $k$ groups with no pairwise
+      difference less than $d$.
+
+      """
+
+Note that in the above example, the first instances of *n* and *k* are
+referring to the input parameters of the function ``stirling``, so they refer
+to Python variables and are therefore formatted as code. The second instances
+of *n* and *k* are talking about mathematical expressions, so they are
+formatted as math.
+
+In general, the rule of thumb is to consider if the variable in question were
+something that rendered differently in code and in math. For example, the Greek
+letter α would be written as ``alpha`` in code and $\alpha$ in LaTeX. The
+reason being that $\alpha$ cannot be used in contexts referring to Python code
+because it is not valid Python, and conversely ``alpha`` would be incorrect in
+a math context because it does not render as the Greek letter (α).
+
+Style Preferences
+-----------------
+
+**Spelling and Punctuation**
+
+All narrative writing in SymPy follows American spelling and punctuation
+standards. For example, “color” is preferred over “colour” and commas should be
+placed inside of quotation marks.
+
+**Examples**
+
+::
+
+  If the ``line_color`` aesthetic is a function of arity 1, then the coloring is
```

Comment:
I think that indentation in RST documents is typically 3 spaces. Maybe we can follow that and also recommend that.

Reply:
How picky is RST about indentation levels? Does it complain if you use 4 spaces instead of 3 in some contexts?

Reply:
Not sure. I don't care what number we choose, but I think it would be good to pick one so that the plain text files are consistent. 3 spaces works with the directive roles nicely:

```
... math::
   a^2
   b^2
```

So that the lines in the directive align with the "m" in math.

Reply:
We're going to add support for double dollars for display math. We just need to merge https://github.com/sympy/sphinx-math-dollar/pull/12 and release the extension.

Reply:
> We're going to add support for double dollars for display math.

I hope that doesn't mean we can no longer use the directives.

Reply:
Both directives still work just fine. Actually it is already enabled in SymPy, even though I only changed one file to use dollars. Everywhere else still uses :math: or .. math:: (or backticks, but those will change).

Reply:
Sounds good @moorepants. It doesn't seem to rendered any differently on the website with 2 versus 3 space indentations, but I can make it all uniform with 3 indentations in the RST file if that is the standard.

Reply:
I just realized that this recommendation could be bad. If you open a python file in your editor you will have the indentation set to 4 spaces. You have to edit the docstrings in that python file, so the indentation is going to be 4 spaces. I don't think many editors would change indentation based on writing docstring or not. In the rst files, your editor might have a setting. Mine is 3 spaces for example. So maybe just make the docstrings 4 spaces since they are in python files. Not sure what to recommend for the rst files, but maybe 4 makes since so that everything is consistent.

Reply:
I was thinking the same thing. In Python files, 4 spaces should be used. Most editors will default to 4 space indentation for Python files, even inside docstrings, so it's best not to fight that. In RST we should use 3 spaces. It matters more in RST because that's when you tend to use the `.. thing` construct that 3 spaces is designed to line up with.

Reply:
Sounds good, I will change the style guide to reflect 3 space indentations under the Narrative Doc guidelines and 4 space indentations under the Docstring guidelines for the python files.